tls/s2n_server_cert_verify.c

```diff
@@ -13,14 +13,151 @@
+    uint8_t unsigned_content_length;
+    uint8_t hash_digest_length;
+    uint16_t signature_size;
+
+    struct s2n_hash_state message_hash;
+    struct s2n_hash_state hash_copy;
+    uint8_t digest_out[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_new(&hash_copy));
+
+    /* Read the algorithm */
+    GUARD(s2n_stuffer_read_uint8(in, &hash_algorithm));
+    GUARD(s2n_stuffer_read_uint8(in, &sign_algorithm));
+    S2N_ERROR_IF(hash_algorithm != conn->secure.conn_hash_alg, S2N_ERR_BAD_MESSAGE);
+    S2N_ERROR_IF(sign_algorithm != conn->secure.conn_sig_alg, S2N_ERR_BAD_MESSAGE);
+
+    /* Verify signature */
+    GUARD(s2n_stuffer_read_uint16(in, &signature_size));
+    S2N_ERROR_IF(signature_size > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+
+    GUARD(s2n_alloc(&signed_content, signature_size));
+    signed_content.size = signature_size;
+    GUARD(s2n_stuffer_read_bytes(in, signed_content.data, signature_size));
+
+    switch (conn->secure.conn_hash_alg) {
+    case S2N_HASH_SHA256:
+        GUARD(s2n_hash_copy(&hash_copy, &conn->handshake.sha256));
+        GUARD(s2n_hash_init(&message_hash, S2N_HASH_SHA256));
+        break;
+    case S2N_HASH_SHA384:
+        GUARD(s2n_hash_copy(&hash_copy, &conn->handshake.sha384));
+        GUARD(s2n_hash_init(&message_hash, S2N_HASH_SHA384));
+        break;
+    case S2N_HASH_NONE:
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA512:
+    case S2N_HASH_MD5_SHA1:
+    case S2N_HASH_SENTINEL:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+        break;
+    }
+
+    GUARD(s2n_hash_digest_size(conn->secure.conn_hash_alg, &hash_digest_length));
+    GUARD(s2n_hash_digest(&hash_copy, digest_out, hash_digest_length));
+
+    uint8_t unsigned_content[(uint8_t) sizeof(digest_out) + s2n_server_cert_verify_header_length()];
+    memcpy(unsigned_content, S2N_CERT_VERIFY_PREFIX, sizeof(S2N_CERT_VERIFY_PREFIX));
+    memcpy(unsigned_content + sizeof(S2N_CERT_VERIFY_PREFIX), S2N_SERVER_CERT_VERIFY_CONTEXT, sizeof(S2N_SERVER_CERT_VERIFY_CONTEXT));
+    memcpy(unsigned_content + s2n_server_cert_verify_header_length(), digest_out, hash_digest_length);
+    unsigned_content_length = hash_digest_length + s2n_server_cert_verify_header_length();
+
+    GUARD(s2n_hash_update(&message_hash, unsigned_content, unsigned_content_length));
+    GUARD(s2n_pkey_verify(&conn->secure.server_public_key, &message_hash, &signed_content));
+
+    return 0;
+}
+
+int s2n_server_write_cert_verify_signature(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    struct s2n_blob signed_content;
+    struct s2n_hash_state message_hash;
+    struct s2n_hash_state hash_copy;
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_new(&hash_copy));
+
+    uint32_t maximum_signature_length = s2n_pkey_size(conn->handshake_params.our_chain_and_key->private_key);
+    GUARD(s2n_alloc(&signed_content, maximum_signature_length));
```

Comment:
if signed_content doesn't need to be allocated on the heap, you can do 
`s2n_stack_blob(signed_content, len, max_len);`